six/modules/c++/six/source/NITFWriteControl.cpp

```diff
@@ -866,11 +866,14 @@ void NITFWriteControl::save(
                 iWriter.setWriteCaching(1);
 
                 nitf::ImageSource iSource;
-                const size_t bandSize = numCols * info.getData()->getNumRows();
+                const NITFSegmentInfo segmentInfo = imageSegments[jj];
+                const size_t bandSize = numCols * segmentInfo.numRows;
 
                 for (size_t chan = 0; chan < numChannels; ++chan)
                 {
-                    nitf::MemorySource ms(imageData[i], bandSize,
+                    nitf::MemorySource ms(imageData[i] +
```

Comment:
The offset needs to also be multiplied by pixelSize.  Rather than doing the pointer arithmetic here, I *think* you can simply pass in `segmentInfo.firstRow * numCols` to the last constructor parameter here (that's set to 0 right now).  But take a look at the underlying code to verify that's true.

To catch this, can you add a test that writes out a multi-image segment SIDD that's 16 bits / pixel?